### ISSUE TITLE
Feature proposal: unsized-media

### ISSUE THREAD

**Issue by loonybear** [WICG/feature-policy#127]
The proposal is to add a new feature, `unsized-media`, which is enabled for all origins by default (default allowlist is `*`)

When disabled, image or video elements (e.g. `<img>`, `<video>`, and `<svg>`), will be using default dimensions (300 x 150) for unspecified dimensions (width and/or height). This prevents relayout of the image or video elements.

**Comment by jpchase** [comment]
Please see the [explainer](https://github.com/WICG/feature-policy/blob/gh-pages/policies/unsized-media.md) for the proposed `unsized-media` policy.

**Comment by annevk** [comment]
So is the goal here to change the behavior of how replaced elements are rendered or only when the replaced element is one of the elements mentioned in the document? E.g., would this affect `::before { content:url(/yolo) }` or `div { content:url(/yolo) }`?

**Comment by annevk** [comment]
How do you get content to jump around with `<svg:image>`?

**Comment by clelland** [comment]
The high level goal is to avoid having replaced elements with an intrinsic size which gets updated after a network request completes. By declaring that image and video elements have *no* intrinsic size, the CSS layout algorithms default to 300x150, exactly as they would for an <iframe>, unless some other constraint forces it to the correct size.

`content:url()` *should* behave the same way, but I suspect it isn't covered in the proposal at all. I'm not sure if Chrome's experimental implementation handles it -- I'll have to check.

**Comment by ojanvafai** [comment]
The goal is specifically to be for media (image, video, svg), not all replaced elements. So, content:url isn't affected, nor are form controls. Is that strange?

It's a good question whether we need this for svg:image. I'm not familiar enough with SVG to know if it's possible to have content jumping due to an svg:image loading.

**Comment by clelland** [comment]
Thanks, @tabatkins! That's the section I spent this morning looking for again.

The algorithms there, in 10.3.2 and 10.6.2, led me to believe that the spec change for this would be to declare that `<img>` and `<video>` elements would have *no intrinsic size* (height, width or ratio), rather than giving them the intrinsic size of the associated image (and updating that when the image becomes available)

**Comment by tabatkins** [comment]
Sure, "no intrinsic size" also works, as far as I can tell.

**Comment by dbaron** [comment]
One other piece is that this should at least be compatible with future mechanisms that allow specifying what the *intrinsic* size or aspect-ratio of an image is.  (These mechanisms would likely be created to solve the problem that there isn't a way to specify the image's intrinsic size without overriding behaviors that honor the aspect ratio.  For example, it would be nice to be able to say something like `<img intrinsic-width=300 intrinsic-height=480 style="width: 100%">` and have the resulting height be the one that keeps the image's aspect ratio with the `100%` width, rather than the resulting height being `480px`.)

In other words, this should only override sizes that would be changed by the process of the image loading and not any others... even if those others are provided by mechanisms that don't exist yet.

**Comment by annevk** [comment]
@clelland so `<img>.naturalWidth`/`naturalHeight`/`naturalRatio` (if we add this one at some point) would return 0/0/0 (or null/null/null if we manage to change)?

This would also affect `createImageBitmap()` and such I think.

Please make sure all such observable aspects have test coverage. (Perhaps making this change by default and running a browser through web-platform-tests you find more aspects that need testing.)

**Comment by ojanvafai** [comment]
@clelland I don't think exposing naturalWidth/naturalHeight at all works for the intended goal. Then people could still write images that are not sized, wait for it to load, read naturalWidth/Height and set it appropriately, no? So I think @annevk's suggestion or something like it is probably what we'd need to do.

@dbaron yup, that makes sense. In fact, in designing this policy, we stumbled across the need for exposing an intrinsicsize property. Proposal is here: https://github.com/ojanvafai/intrinsicsize-attribute. Feedback very welcome.

**Comment by AmeliaBR** [comment]
(I was directed here to weigh in on the interaction with SVG's `<image>`; not sure if the proposal is still active or if it's waiting on better responsive aspect ratio control in markup.)

My main question: What is the actual goal of restricting this feature? Because that should influence exactly which factors are limited.

If the goal is only to prevent layout from jumping around based on content from untrusted domains, then there's no need to obfuscate the DOM properties. Also, there's no need to worry about `<svg:image>`, since it's size cannot affect the layout of other elements even when it is auto-sized. (In SVG 2, anyway. If we ever add automatic calculation of the `<svg>` element's aspect ratio based on its graphical contents, then there would be an indirect effect from an auto-sized image.)

If the goal is to prevent scripts on this domain from being able to exploit the legacy cross-origin security holes that come from being able to read an image's dimensions and confirm that it has correctly loaded, then the more extreme restrictions seem appropriate.

**Comment by clelland** [comment]
Hi @AmeliaBR,
I don't believe that this was originally intended as a security feature at all; the original intention was to stabilize layout by requiring all images to have a size that can be determined by markup / style alone. (Similar to the way that [&lt;amp-img&gt;](https://www.ampproject.org/docs/reference/components/amp-img) requires `width` and `height` to be set)

(And the proposal is still active; this is implemented in Chrome behind a flag, and there's a demo page up at https://feature-policy-demos.appspot.com/unsized-media.html)

**Comment by tabatkins** [comment]
Right, this is purely a "mess me up if I write my page badly, in a way that makes the browser's job easier" feature, like the `contain` property. No security properties at all.
